forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/as/api/authorisation/redirect/AuthorisationApiController.java

```diff
@@ -54,95 +56,124 @@
+    public AuthorisationApiController(HeadLessAuthorisationService headLessAuthorisationService,
+                                      CryptoApiClient cryptoApiClient, TppStoreService tppStoreService,
+                                      @Value("${am.cookie.name}") String cookieName,
+                                      @Value("${as.headless.always-enable}") boolean isHeadlessAlwaysEnabled,
+                                      AMGatewayService amGatewayService, JwtOverridingService jwtOverridingService,
+                                      TokenUsageService tokenUsageService, DiscoveryConfig discoveryConfig) {
+        this.headLessAuthorisationService = headLessAuthorisationService;
+        this.cryptoApiClient = cryptoApiClient;
+        this.tppStoreService = tppStoreService;
+        this.cookieName = cookieName;
+        this.isHeadlessAlwaysEnabled = isHeadlessAlwaysEnabled;
+        this.amGatewayService = amGatewayService;
+        this.jwtOverridingService = jwtOverridingService;
+        this.tokenUsageService = tokenUsageService;
+        this.discoveryConfig = discoveryConfig;
+    }
 
-            @ApiParam(value = "Default username for the headless authentication")
-            @RequestHeader(value = "X_HEADLESS_AUTH_ENABLED", required = false, defaultValue = "false") boolean isHeadlessEnabled,
+    private static boolean hasQueryParamIdToken(ResponseEntity responseEntity) {
+        if (hasNoLocationHeader(responseEntity)) return false;
+        if (responseEntity.getHeaders().getLocation().getQuery() == null) {
+            return false;
+        }
+        return responseEntity.getHeaders().getLocation().getQuery().contains("error_description");
+    }
 
-            @ApiParam(value = "Default username for the headless authentication")
-            @RequestHeader(value = "X_HEADLESS_AUTH_USERNAME", required = false, defaultValue = "") String username,
 
-            @ApiParam(value = "Default username password for the headless authentication")
-            @RequestHeader(value = "X_HEADLESS_AUTH_PASSWORD", required = false, defaultValue = "") String password,
+    private static boolean hasFragmentIdToken(ResponseEntity responseEntity) {
+        if (hasNoLocationHeader(responseEntity)) return false;
+        if (responseEntity.getHeaders().getLocation().getFragment() == null) {
+            return false;
+        }
 
-            @ApiParam(value = "Cookie containing the user session", required = false)
-            @CookieValue(value = "${am.cookie.name}", required = false) String ssoToken,
+        if (!responseEntity.getHeaders().getLocation().getFragment().contains("id_token")) {
+            log.debug("Fragment {} doesn't contains id_token", responseEntity.getHeaders().getLocation().getFragment());
+            return false;
+        }
+        return true;
+    }
 
-            @RequestBody(required = false) MultiValueMap body,
+    private static boolean hasNoLocationHeader(ResponseEntity responseEntity) {
```

Comment:
Nice one - I considered mentioning it could do with this method to reduce the duplication 👍🏻